seeds: add signet/testnet4, update makeseeds regex, minblocks, fixed seeds

jonatack opened the pull request:
In `makeseeds.py`:
- fix the user agent regex (by laanwj)
- fix the I2P column header spacing
- update the regex (it was also not updated for the previous release)
- update `MIN_BLOCKS` (4320 blocks/month * ~6.5 months) and add README documentation to remember to update it
- further robustness/standardness/consistency improvements to the regexes (by l0rinc) 

Add signet and testnet4 seeds to the README and to `generate-seeds.py`

Update the fixed seeds in `src/chainparamsseeds.h`

In `kernel/chainparams.cpp`:
- add signet fixed seeds if default network
- remove hardcoded signet seeds

Update `contrib/seeds/.gitignore` with signet and testnet4

The previous 2 seeds updates were https://github.com/bitcoin/bitcoin/pull/30008 and https://github.com/bitcoin/bitcoin/pull/30695.

mainnet:
```
$ contrib/seeds$ python3 makeseeds.py -a asmap-filled.dat -s seeds_main.txt > nodes_main.txt

Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass                                               
 17252   3630  21079   3095     12 Initial
 17252   3630  21079   3095     12 Skip entries with invalid address
  8444   1742  14607   2330     10 After removing duplicates
  8194   1691  14321   2102     10 Enforce minimal number of blocks
  7838   1578  14321   2102     10 Require service bit 1
  6802   1326  14321   2102     10 Require minimum uptime
  6762   1321  14320   2102     10 Require a known and recent user agent
  6762   1321  14320   2102     10 Filter out hosts with multiple bitcoin ports
   512    485    512    512     10 Look up ASNs and limit results per ASN and per net
```

signet:
```
$ contrib/seeds$ python3 makeseeds.py -a asmap-filled.dat -s seeds_signet.txt -m 237800 > nodes_signet.txt


Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass                                               
   110     47     63      9      4 Initial
   110     47     63      9      4 Skip entries with invalid address
   110     47     63      9      4 After removing duplicates
    83     31     58      9      4 Enforce minimal number of blocks
    83     31     58      9      4 Require service bit 1
    83     31     57      9      4 Require minimum uptime
    83     31     57      9      4 Require a known and recent user agent
    83     31     57      7      4 Filter out hosts with multiple bitcoin ports
    42     30     57      7      4 Look up ASNs and limit results per ASN and per net
```

testnet:
```
$ contrib/seeds$ python3 makeseeds.py -a asmap-filled.dat -s seeds_test.txt > nodes_test.txt

Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass                                               
   204     73     96     11      5 Initial
   204     73     96     11      5 Skip entries with invalid address
   204     73     96     11      5 After removing duplicates
   204     73     96     11      5 Enforce minimal number of blocks
   204     73     96     11      5 Require service bit 1
   195     69     87      9      5 Require minimum uptime
   193     69     87      9      5 Require a known and recent user agent
   193     69     87      9      5 Filter out hosts with multiple bitcoin ports
    79     39     87      9      5 Look up ASNs and limit results per ASN and per net
```

testnet4
```
$ contrib/seeds$ python3 makeseeds.py -a asmap-filled.dat -s seeds_testnet4.txt -m 72600 > nodes_testnet4.txt

Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass                                               
   149    115     69     11      4 Initial
   149    115     69     11      4 Skip entries with invalid address
   149    115     69     11      4 After removing duplicates
   104     75     52      7      4 Enforce minimal number of blocks
   104     75     52      7      4 Require service bit 1
   100     73     51      7      4 Require minimum uptime
   100     73     51      7      4 Require a known and recent user agent
   100     73     51      7      4 Filter out hosts with multiple bitcoin ports
    43     46     51      7      4 Look up ASNs and limit results per ASN and per net
```

jonatack committed:
makeseeds: update user agent regex

laanwj commented on contrib/seeds/makeseeds.py:
```diff
@@ -39,16 +39,16 @@
     r"0.17.(0|0.1|1|2|99)|"
     r"0.18.(0|1|99)|"
-    r"0.19.(0|1|2|99)|"
+    r"0.19.(0|1|2|99).(0|1)|"
```
The regexp doesn't match the entire string until the end, so "0.19.0.1" and such already match. This change will only make it that "0.19.1" no longer matches.

jonatack replied:
Oops -- done, thank you!

laanwj commented on contrib/seeds/makeseeds.py:
```diff
@@ -39,16 +39,16 @@
-    r"28.(0|99).0|"
+    r"26.(0|1|2|99).0|"
+    r"27.(0|1|2|99).0|"
+    r"28.(0|1|99).0|"
```
We shouldn't have `|` at the end of the last clause, as this will make it match the empty string too (so effectively everything starting with `Satoshi:` matches).

jonatack replied:
> We shouldn't have `|` at the end of the last clause, as this will make it match the empty string too (so effectively everything starting with `Satoshi:` matches).

Good catch. It looks like this would mean that the regex has been doing this since a8ec9eede4c745c6b6fd76974816ffad8034129a one year ago?

laanwj replied:
Yes, i didn't notice before ! It's too easy to make this mistake.

It also makes me wonder whether this is a useful filter at all, or whether to replace it with a simpler one that extracts the version number, followed by some minimum/maximum bounds check. But that's not something that needs to be addressed in this PR.

laanwj replied:
Thanks!

laanwj committed:
makeseeds: fix incorrect regex

We shouldn't have | at the end of the last clause, as this will make it match
the empty string too (so effectively everything starting with Satoshi: matches).

While doing this, put the | at the beginning of every line of regex rather than
the end, to make it easier to update in the future without accidentally running
into this problem again.

jonatack committed:
makeseeds: align I2P column header

jonatack commented:
Updated to take @laanwj's review feedback and re-ran the seeds scripts.

laanwj commented:
Code review ACK b2dc4458ac86c0a633ca3eb87827ac2731993f97 on the script changes

mainnet
```
Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass
 17173   3638  21303   3116     12 Initial
 17173   3638  21303   3116     12 Skip entries with invalid address
  8378   1755  14661   2343     10 After removing duplicates
  8337   1736  14656   2343     10 Enforce minimal number of blocks
  7986   1612  14654   2343     10 Require service bit 1
  6963   1365  14644   2343     10 Require minimum uptime
  6917   1359  14643   2343     10 Require a known and recent user agent
  6917   1359  14643   2343     10 Filter out hosts with multiple bitcoin ports
   512    502    512    512     10 Look up ASNs and limit results per ASN and per net
```

testnet
```
Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion    I2P  CJDNS Pass
   204     74     95     11      5 Initial
   204     74     95     11      5 Skip entries with invalid address
   204     74     95     11      5 After removing duplicates
   204     74     95     11      5 Enforce minimal number of blocks
   204     74     95     11      5 Require service bit 1
   197     68     87      9      5 Require minimum uptime
   195     68     87      9      5 Require a known and recent user agent
   195     68     87      9      5 Filter out hosts with multiple bitcoin ports
    78     38     87      9      5 Look up ASNs and limit results per ASN and per net
```

i have re-run the steps in an attempt to check the output, but the lack of determinism makes it hard to say anything. All that could be concluded is that the diff is smaller from the previous state to the current one, to the one between the current one and my re-run:

```bash
$ git show HEAD --shortstat nodes_main.txt
commit b2dc4458ac86c0a633ca3eb87827ac2731993f97 (HEAD)
 1 file changed, 1618 insertions(+), 1627 deletions(-)
$ git diff --shortstat nodes_main.txt
 1 file changed, 419 insertions(+), 418 deletions(-)

$ git show HEAD --shortstat nodes_test.txt
commit b2dc4458ac86c0a633ca3eb87827ac2731993f97 (HEAD)
 1 file changed, 128 insertions(+), 222 deletions(-)
$ git diff --shortstat nodes_test.txt
(no output)
```
The testnet output does match 100%.

jonatack commented:
Yes, for the avoidance of doubt, I ran the script with no manual curation (as done previously from v21 to v27).

achow101 commented:
Could also do testnet4, my seeder publishes a seeds.txt.gz at https://testnet4.achownodes.xyz/seeds.txt.gz

achow101 commented on contrib/seeds/makeseeds.py:
```diff
@@ -48,7 +48,7 @@
     r"26.(0|1|99).0|"
     r"27.(0|1|99).0|"
-    r"28.(0|99).0|"
+    r"28.(0|99).0"
```
In 6597e87666daa7fabdf310e9ad39fc8a7c81ceea "makeseeds: fix incorrect regex"

Perhaps put the `|` in the front of every regex rather than the end, to make it easier to update in the future and not accidentally run into this problem again.

jonatack replied:
Updating to take this feedback and https://github.com/bitcoin/bitcoin/pull/31960#issuecomment-2695738356, thanks.

jonatack replied:
Done

jonatack committed:
seeds: add signet seeds

jonatack committed:
seeds: add testnet4 seeds

jonatack commented:
Should `MIN_BLOCKS = 840000` be updated? It was last set in 5bab3175a663610070c1000dd4211a58490e5023.

jonatack committed:
generate-seeds: update and add signet

jonatack committed:
makeseeds: update MIN_BLOCKS, add reminder to README

jonatack commented:
> Could also do testnet4, my seeder publishes a seeds.txt.gz at [testnet4.achownodes.xyz/seeds.txt.gz](https://testnet4.achownodes.xyz/seeds.txt.gz). And also signet: [signet.achownodes.xyz/seeds.txt.gz](https://signet.achownodes.xyz/seeds.txt.gz)

Done. Also updated `MIN_BLOCKS` and added a reminder to the README.

l0rinc reviewed (changes requested):
I have only reviewed the regexes, please see my suggestions.

<details>
<summary>Regexes</summary>

```python
PATTERN_IPV4 = re.compile(r"^(([0-2]?\d{1,2})\.([0-2]?\d{1,2})\.([0-2]?\d{1,2})\.([0-2]?\d{1,2})):(\d{1,5})$")
PATTERN_IPV6 = re.compile(r"^\[([\da-f:]+)]:(\d{1,5})$", re.IGNORECASE)
PATTERN_ONION = re.compile(r"^([a-z2-7]{56}\.onion):(\d{1,5})$")
PATTERN_I2P = re.compile(r"^([a-z2-7]{52}\.b32\.i2p):(\d{1,5})$")
PATTERN_AGENT = re.compile(
    r"^/Satoshi:("
    r"0\.14\.([0-3]|99)"
    r"|0\.15\.([0-2]|99)"
    r"|0\.16\.([0-3]|99)"
    r"|0\.17\.(0\.1|[0-2]|99)"
    r"|0\.18\.([01]|99)"
    r"|0\.19\.([0-2]|99)"
    r"|0\.20\.([0-2]|99)"
    r"|0\.21\.([0-2]|99)"
    r"|22\.([01]|99)\.0"
    r"|23\.([01]|99)\.0"
    r"|24\.([0-2]|99)\.[01]"
    r"|25\.([0-2]|99)\.0"
    r"|26\.([01]|99)\.0"
    r"|27\.([01]|99)\.0"
    r"|28\.(0|99)\.0"
    r")")
```

</details>

achow101 commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
+    r"|26.(0|1|2|99).0"
+    r"|27.(0|1|2|99).0"
+    r"|28.(0|1|99).0"
     r")")
```
Prefixes, the user agent may contain a comment at the end which we can ignore.

achow101 commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
+    r"0.14.(0|1|2|3|99)"
+    r"|0.15.(0|1|2|99)"
+    r"|0.16.(0|1|2|3|99)"
+    r"|0.17.(0|0.1|1|2|99)"
```
I don't think this needs compression, explicitly listing the minor versions is easier to read IMO.

achow101 commented:
While src/chainparamsseeds.h now has `chainparams_seed_signet`, it is not used anywhere. `SigNetParams` does not set `vFixedSeeds` as the other ones do. It probably should, but only when using the default testnet, as we do for the DNS seeds.

jonatack committed:
chainparams: add signet fixed seeds if default network

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
-    r"26.(0|1|99).0|"
-    r"27.(0|1|99).0|"
-    r"28.(0|99).0|"
+    r"0.14.(0|1|2|3|99)"
```
Hi @l0rinc, this diff, if I understand you correctly?

```diff
 PATTERN_AGENT = re.compile(
     r"^/Satoshi:("
-    r"0.14.(0|1|2|3|99)"
-    r"|0.15.(0|1|2|99)"
-    r"|0.16.(0|1|2|3|99)"
-    r"|0.17.(0|0.1|1|2|99)"
-    r"|0.18.(0|1|99)"
-    r"|0.19.(0|1|2|99)"
-    r"|0.20.(0|1|2|99)"
-    r"|0.21.(0|1|2|99)"
-    r"|22.(0|1|99).0"
-    r"|23.(0|1|2|99).0"
-    r"|24.(0|1|2|99).(0|1)"
-    r"|25.(0|1|2|99).0"
-    r"|26.(0|1|2|99).0"
-    r"|27.(0|1|2|99).0"
-    r"|28.(0|1|99).0"
+    r"0\.14\.([0-3]|99)"
+    r"|0\.15\.(0|1|2|99)"
+    r"|0\.16\.(0|1|2|3|99)"
+    r"|0\.17\.(0|0.1|1|2|99)"
+    r"|0\.18\.(0|1|99)"
+    r"|0\.19\.(0|1|2|99)"
+    r"|0\.20\.(0|1|2|99)"
+    r"|0\.21\.(0|1|2|99)"
+    r"|22\.(0|1|99)\.0"
+    r"|23\.(0|1|2|99)\.0"
+    r"|24\.(0|1|2|99)\.(0|1)"
+    r"|25\.(0|1|2|99)\.0"
+    r"|26\.(0|1|2|99)\.0"
+    r"|27\.(0|1|2|99)\.0"
+    r"|28\.(0|1|99)\.0"
     r")")
```

l0rinc committed:
makeseeds: regex improvements

jonatack committed:
chainparams: remove hardcoded signet seeds

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
+MIN_BLOCKS = 868000
 
 PATTERN_IPV4 = re.compile(r"^((\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})):(\d+)$")
 PATTERN_IPV6 = re.compile(r"^\[([0-9a-z:]+)\]:(\d+)$")
```
done

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
-MIN_BLOCKS = 840000
+MIN_BLOCKS = 868000
 
 PATTERN_IPV4 = re.compile(r"^((\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})):(\d+)$")
```
done

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
 PATTERN_IPV4 = re.compile(r"^((\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})):(\d+)$")
 PATTERN_IPV6 = re.compile(r"^\[([0-9a-z:]+)\]:(\d+)$")
 PATTERN_ONION = re.compile(r"^([a-z2-7]{56}\.onion):(\d+)$")
 PATTERN_I2P = re.compile(r"^([a-z2-7]{52}\.b32.i2p):(\d+)$")
```
done

jonatack committed:
seeds: update fixed dns seeds

jonatack committed:
seeds: update .gitignore with signet and testnet4

jonatack commented:
Thanks for the reviews @l0rinc and @achow101. Updated to take @achow101's suggestion, added a commit by @l0rinc with his review feedback (makeseeds output appears unchanged by the robustness/consistency changes to the regexes), updated `.gitignore` in `/contrib/seeds`, and re-ran the scripts.

jonatack commented:
Re-pushed to add a missing escape to the regexes commit.

achow101 commented:
Needs rebase

jonatack commented:
Rebased

achow101 commented:
ACK 89fd9a4a87d772798671da18538e45f0272365e4

l0rinc commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
-PATTERN_IPV6 = re.compile(r"^\[([0-9a-z:]+)\]:(\d+)$")
+PATTERN_IPV4 = re.compile(r"^(([0-2]?\d{1,2})\.([0-2]?\d{1,2})\.([0-2]?\d{1,2})\.([0-2]?\d{1,2})):(\d{1,5})$")
+PATTERN_IPV6 = re.compile(r"^\[([\da-f:]+)]:(\d{1,5})$", re.IGNORECASE)
 PATTERN_ONION = re.compile(r"^([a-z2-7]{56}\.onion):(\d+)$")
```
For consistency this can also be:
```suggestion
PATTERN_ONION = re.compile(r"^([a-z2-7]{56}\.onion):(\d{1,5})$")
```

jonatack replied:
Will do if I need to re-run the scripts and re-push.

l0rinc commented on src/kernel/chainparams.cpp:
```diff
@@ -417,17 +417,15 @@ class SigNetParams : public CChainParams {
 
         if (!options.challenge) {
             bin = "512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae"_hex_v_u8;
+            vFixedSeeds = std::vector<uint8_t>(std::begin(chainparams_seed_signet), std::end(chainparams_seed_signet));
```
1) the `chainparams: add signet fixed seeds if default network` commit should go after `seeds: update fixed dns seeds` which introduces `chainparams_seed_signet`
2) Now that we're pre-clearing it at the beginning, it may be slightly simpler to do:
```suggestion
            vFixedSeeds.assign(std::begin(chainparams_seed_signet), std::end(chainparams_seed_signet));

```

l0rinc replied:
I see, what do you think about 2)?

l0rinc commented on contrib/seeds/makeseeds.py:
```diff
@@ -26,29 +26,29 @@
-    r"24.(0|1|2|99).(0|1)|"
-    r"25.(0|1|2|99).0|"
-    r"26.(0|1|99).0|"
-    r"27.(0|1|99).0|"
-    r"28.(0|99).0|"
+    r"0\.14\.(0|1|2|3|99)"
+    r"|0\.15\.(0|1|2|99)"
+    r"|0\.16\.(0|1|2|3|99)"
+    r"|0\.17\.(0|0\.1|1|2|99)"
+    r"|0\.18\.(0|1|99)"
+    r"|0\.19\.(0|1|2|99)"
+    r"|0\.20\.(0|1|2|99)"
+    r"|0\.21\.(0|1|2|99)"
+    r"|22\.(0|1|99)\.0"
+    r"|23\.(0|1|2|99)\.0"
+    r"|24\.(0|1|2|99)\.(0|1)"
+    r"|25\.(0|1|2|99)\.0"
+    r"|26\.(0|1|2|99)\.0"
+    r"|27\.(0|1|2|99)\.0"
+    r"|28\.(0|1|99)\.0"
     r")")
```
> Those are for additional releases

Hmmm, if I get the list of every version via `git fetch upstream --prune --prune-tags --force && git tag --sort=creatordate | sed -n 's#^v##p'`, I'm getting *319 tags*, many of which (including the latest ones) aren't currently matched by `PATTERN_AGENT`, see https://www.online-python.com/3iU2QMKu97

Can we generalize the agent pattern to something like this instead?
```python
PATTERN_AGENT = re.compile(r"^/Satoshi:(\d+(\.\d+){0,3}[-\w]*)$")
```

jonatack replied:
> This pattern is specifically not generalized because we want to restrict to at least post-segwit versions, hence beginning at 0.14.0. We might even want to go further and restrict to post-taproot versions, although the benefit there is much less significant.
> 
> Furthermore, while the tag is similar to the user agent, they do not match exactly. Specifically, since the version style change in 22.0, the tag is typically MAJOR.MINOR, but the user agent is always MAJOR.MINOR.PATCH. So 22.0 is 22.0.0 in the user agent. Release candidates additionally produce user agents without the rc suffix. Hence your test script is not representative of what the makeseeds.py actually tests.
> 
> Lastly, we explicitly list out the possible versions in order to be more confident that the included seed nodes are running actual Bitcoin Core releases, and also master branch builds.

Good explanation. As a follow-up might be worth documenting in `makeseeds.py` and/or the readme.

jonatack replied:
Here is a python script that sanity checks the regex in this PR. Only the last entry "24.1rc1" doesn't match. Changing entries to, say, 24.3.0 also don't match.

<details><summary>regex-test.py</summary><p>

```python3
#!/usr/bin/env python3
import re

TAGS = """
0.14.0
0.14.1
0.14.2
0.14.3
0.15.0
0.15.0.1
0.15.1
0.16.0
0.16.1
0.16.2
0.16.3
0.15.2
0.14.3
0.17.0
0.17.0.1
0.17.1
0.18.0
0.18.1
0.17.2
0.19.0
0.19.0.1
0.19.1
0.20.0
0.20.1
0.19.2
0.21.0
0.21.1
22.0.0
0.21.2
0.20.2
23.0.0
24.0.0
22.1.0
24.0.1
23.1.0
24.1.0
23.2.0
25.0.0-final
25.1.0
24.2.0
26.0.0
26.1.0
25.2.0
25.1.0(FutureBit-Apollo-Node)
27.0.0
27.1.0
26.2.0
28.0.0
27.2.0
28.1.0
28.99.0
0.20.0-final
24.1rc1
""".strip().splitlines()

# pattern = re.compile(r'^\d+(\.\d+){0,3}[-\w]*$')
pattern = re.compile(
    r"^/Satoshi:("
    r"0\.14\.(0|1|2|3|99)"
    r"|0\.15\.(0|1|2|99)"
    r"|0\.16\.(0|1|2|3|99)"
    r"|0\.17\.(0|0\.1|1|2|99)"
    r"|0\.18\.(0|1|99)"
    r"|0\.19\.(0|1|2|99)"
    r"|0\.20\.(0|1|2|99)"
    r"|0\.21\.(0|1|2|99)"
    r"|22\.(0|1|99)\.0"
    r"|23\.(0|1|2|99)\.0"
    r"|24\.(0|1|2|99)\.(0|1)"
    r"|25\.(0|1|2|99)\.0"
    r"|26\.(0|1|2|99)\.0"
    r"|27\.(0|1|2|99)\.0"
    r"|28\.(0|1|99)\.0"
    r")")

for tag in TAGS:
    if not pattern.match(f"/Satoshi:{tag}/"):
        print(f"Tag did NOT match: {tag}")
        exit(1)

print("All tags matched successfully.")
```
</p></details>

laanwj replied:
> I don't see why that would be better than this approach.

The only possible advantage i can think of to use a straight-up list and a fixed regexp, would be that it is less prone to subtle errors in the regexp like https://github.com/bitcoin/bitcoin/pull/31960#pullrequestreview-2654030620 .

achow101 commented:
ACK f0b659716bd455dca02053df573d888b5a115aa4

laanwj reviewed (approved):
re-ACK f0b659716bd455dca02053df573d888b5a115aa4
